File: prob10/scheduler.py
# Problem 10
#This problem was asked by Apple.

#Implement a job scheduler which takes in a 
# function f and an integer n, and calls f after 
# n milliseconds.
import time
import threading

def schedule(f, n, *args):
    #thread = threading.Thread(target=f, args=("arg1", n, ". All done!", ))
    #thread.start()
    print("Scheduler started")
    time.sleep(n)
    #f("arg1", "arg2", "All done!")
    f()
    print("Scheduler finished")

def func(args=""):
    print("Function called", end="")
#    for arg in args:
#        print(arg, end=" ")
#    print()

def main():
    time_sleep = 3
    thread = threading.Thread(target=schedule, args=(func, time_sleep, ". Good Job!", ))
    thread.start()
    print("Hi")

if __name__ == "__main__":
    main()

File: prob10/scheduler.go
package main

/* Problem 10
 * This problem was asked by Apple.

 * Implement a job scheduler which takes in a
 * function f and an integer n, and calls f after
 * n milliseconds
 */
import (
	"fmt"
	"sync"
	"time"
)

var wg sync.WaitGroup
var wg1 sync.WaitGroup

func main() {
	fmt.Println("Hello")
	wg.Add(1)
	go scheduler(myfunc, 800)

	wg.Wait()
	fmt.Println("All tasks finished")
}

func myfunc() {
	defer wg1.Done()
	fmt.Println("Function started")
}

func scheduler(f func(), n int) {
	defer wg.Done()

	fmt.Println("Scheduler started")
	wg1.Add(1)

	time.Sleep(time.Duration(n) * time.Millisecond)
	go f()
	wg1.Wait()
	fmt.Println("Scheduler finished")
}
